File: mekaneck/virtualbrain/orchestrator.py
"""
VirtualBrainValidationSuite: Orchestrates complete validation.

Follows blindhorse pattern for validation orchestration.
"""

import time
from pathlib import Path
from typing import Dict, List, Optional

from .validators import (
    PartitionValidator,
    TripleEquivalenceValidator,
    KuramotoValidator,
    ConsciousnessValidator,
)
from .utils import save_json


class VirtualBrainValidationSuite:
    """
    Orchestrates complete Virtual Brain validation.

    Runs all validators in sequence, collects results,
    and produces comprehensive report.
    """

    def __init__(self, output_dir: Path = Path("results")):
        self.output_dir = output_dir
        self.output_dir.mkdir(parents=True, exist_ok=True)

        self.results: Dict = {}
        self.start_time = None
        self.end_time = None

    def run_complete_validation(
        self,
        skip_validators: Optional[List[str]] = None,
    ) -> Dict:
        """
        Run complete validation suite.

        Args:
            skip_validators: List of validator names to skip

        Returns:
            Complete results dictionary
        """
        skip_validators = skip_validators or []

        print("\n" + "=" * 70)
        print(" " * 15 + "VIRTUAL BRAIN COMPUTING FRAMEWORK")
        print(" " * 20 + "VALIDATION SUITE")
        print("=" * 70)

        self.start_time = time.time()

        # Define validators in order
        validators = [
            ("partition", PartitionValidator, "Partition Coordinates"),
            ("triple_equivalence", TripleEquivalenceValidator, "Triple Equivalence"),
            ("kuramoto", KuramotoValidator, "Kuramoto Dynamics"),
            ("consciousness", ConsciousnessValidator, "Consciousness Equations"),
        ]

        # Run each validator
        for idx, (name, ValidatorClass, description) in enumerate(validators, 1):
            if name in skip_validators:
                print(f"\n[SKIP] Validator {idx}/{len(validators)}: {description}")
                continue

            try:
                print(f"\n" + "=" * 70)
                print(f"VALIDATOR {idx}/{len(validators)}: {description}")
                print("=" * 70)

                validator = ValidatorClass(output_dir=self.output_dir / name)
                self.results[name] = validator.run_validation()

            except Exception as e:
                print(f"ERROR in {description}: {e}")
                self.results[name] = {"error": str(e)}

        self.end_time = time.time()

        # Generate report
        self._generate_report()

        # Save complete results
        self._save_complete_results()

        return self.results

    def _generate_report(self):
        """Generate comprehensive validation report."""
        print("\n" + "=" * 70)
        print(" " * 20 + "VALIDATION REPORT")
        print("=" * 70)

        total_time = self.end_time - self.start_time

        total_validators = len(self.results)
        successful = sum(1 for r in self.results.values() if "error" not in r)

        print(f"\nExecution Summary:")
        print(f"  Total validators: {total_validators}")
        print(f"  Successful: {successful}")
        print(f"  Failed: {total_validators - successful}")
        print(f"  Total time: {total_time:.2f} seconds")

        # Count claims
        total_claims = 0
        validated_claims = 0

        for validator_name, results in self.results.items():
            if "claims_validated" in results:
                claims = results["claims_validated"]
                total_claims += len(claims)
                validated_claims += sum(1 for c in claims.values() if c)

        validation_rate = validated_claims / total_claims if total_claims > 0 else 0

        print(f"\nValidation Summary:")
        print(f"  Total claims: {total_claims}")
        print(f"  Validated: {validated_claims}")
        print(f"  Validation rate: {validation_rate:.1%}")

        if validation_rate >= 0.9:
            status = "[OK] EXCELLENT - Framework validated"
        elif validation_rate >= 0.8:
            status = "[OK] GOOD - Most claims validated"
        elif validation_rate >= 0.7:
            status = "[WARN] ACCEPTABLE - Needs review"
        else:
            status = "[FAIL] NEEDS WORK - Significant issues"

        print(f"\nOverall Status: {status}")
        print("=" * 70)

    def _save_complete_results(self):
        """Save complete results to JSON."""
        output_file = self.output_dir / "complete_validation_results.json"

        results_with_metadata = {
            "metadata": {
                "suite": "Virtual Brain Computing Framework Validation",
                "version": "0.1.0",
                "timestamp": time.strftime("%Y-%m-%d %H:%M:%S"),
                "total_time_seconds": float(self.end_time - self.start_time),
            },
            "results": self.results,
        }

        save_json(results_with_metadata, output_file)
        print(f"\n[OK] Complete results saved to: {output_file}")

File: virtual_brain/crates/vb-validators/src/orchestrator.rs
//! Orchestrator: Runs all validators and produces comprehensive report.

use crate::base::{ValidationResult, Validator};
use crate::consciousness_validator::ConsciousnessValidator;
use crate::kuramoto_validator::KuramotoValidator;
use crate::partition_validator::PartitionValidator;
use anyhow::Result;
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::time::Instant;

/// Validation suite that runs all validators.
pub struct ValidationSuite {
    output_dir: PathBuf,
    results: HashMap<String, ValidationResult>,
    start_time: Option<Instant>,
    end_time: Option<Instant>,
}

impl ValidationSuite {
    /// Create new validation suite.
    pub fn new(output_dir: impl AsRef<Path>) -> Self {
        Self {
            output_dir: output_dir.as_ref().to_path_buf(),
            results: HashMap::new(),
            start_time: None,
            end_time: None,
        }
    }

    /// Run complete validation.
    pub fn run_complete_validation(&mut self, skip_validators: &[&str]) -> Result<&HashMap<String, ValidationResult>> {
        println!("\n{}", "=".repeat(70));
        println!("{:^70}", "VIRTUAL BRAIN COMPUTING FRAMEWORK");
        println!("{:^70}", "VALIDATION SUITE (Rust)");
        println!("{}", "=".repeat(70));

        self.start_time = Some(Instant::now());

        // Define validators
        let validators: Vec<(&str, Box<dyn Validator>)> = vec![
            (
                "partition",
                Box::new(PartitionValidator::new(10, self.output_dir.join("partition"))),
            ),
            (
                "kuramoto",
                Box::new(KuramotoValidator::new(100, self.output_dir.join("kuramoto"))),
            ),
            (
                "consciousness",
                Box::new(ConsciousnessValidator::new(self.output_dir.join("consciousness"))),
            ),
        ];

        let total = validators.len();

        for (idx, (name, mut validator)) in validators.into_iter().enumerate() {
            if skip_validators.contains(&name) {
                println!("\n[SKIP] Validator {}/{}: {}", idx + 1, total, name);
                continue;
            }

            println!("\n{}", "=".repeat(70));
            println!("VALIDATOR {}/{}: {}", idx + 1, total, name.to_uppercase());
            println!("{}", "=".repeat(70));

            match validator.run_validation() {
                Ok(result) => {
                    self.results.insert(name.to_string(), result);
                }
                Err(e) => {
                    println!("ERROR in {}: {}", name, e);
                }
            }
        }

        self.end_time = Some(Instant::now());

        // Generate report
        self.generate_report();

        // Save complete results
        self.save_complete_results()?;

        Ok(&self.results)
    }

    fn generate_report(&self) {
        println!("\n{}", "=".repeat(70));
        println!("{:^70}", "VALIDATION REPORT");
        println!("{}", "=".repeat(70));

        let total_time = self
            .end_time
            .and_then(|e| self.start_time.map(|s| e.duration_since(s)))
            .map(|d| d.as_secs_f64())
            .unwrap_or(0.0);

        let total_validators = self.results.len();

        println!("\nExecution Summary:");
        println!("  Total validators: {}", total_validators);
        println!("  Total time: {:.2} seconds", total_time);

        // Count claims
        let mut total_claims = 0;
        let mut validated_claims = 0;

        for result in self.results.values() {
            total_claims += result.total_claims();
            validated_claims += result.validated_count();
        }

        let validation_rate = if total_claims > 0 {
            validated_claims as f64 / total_claims as f64
        } else {
            0.0
        };

        println!("\nValidation Summary:");
        println!("  Total claims: {}", total_claims);
        println!("  Validated: {}", validated_claims);
        println!("  Validation rate: {:.1}%", validation_rate * 100.0);

        let status = if validation_rate >= 0.9 {
            "[OK] EXCELLENT - Framework validated"
        } else if validation_rate >= 0.8 {
            "[OK] GOOD - Most claims validated"
        } else if validation_rate >= 0.7 {
            "[WARN] ACCEPTABLE - Needs review"
        } else {
            "[FAIL] NEEDS WORK - Significant issues"
        };

        println!("\nOverall Status: {}", status);
        println!("{}", "=".repeat(70));
    }

    fn save_complete_results(&self) -> Result<()> {
        std::fs::create_dir_all(&self.output_dir)?;

        let output_file = self.output_dir.join("complete_validation_results.json");

        let total_time = self
            .end_time
            .and_then(|e| self.start_time.map(|s| e.duration_since(s)))
            .map(|d| d.as_secs_f64())
            .unwrap_or(0.0);

        let results_with_metadata = serde_json::json!({
            "metadata": {
                "suite": "Virtual Brain Computing Framework Validation (Rust)",
                "version": "0.1.0",
                "timestamp": chrono::Local::now().format("%Y-%m-%d %H:%M:%S").to_string(),
                "total_time_seconds": total_time,
            },
            "results": self.results,
        });

        let json = serde_json::to_string_pretty(&results_with_metadata)?;
        std::fs::write(&output_file, json)?;

        println!("\n[OK] Complete results saved to: {}", output_file.display());

        Ok(())
    }
}
